src/qibo/validation/gate_set_tomography.py

```diff
@@ -0,0 +1,846 @@
+        measurement outcomes for the Pauli Transfer Matrix notation. The circuit
+        should already have noise models implemented, if any, prior to using this
+        function.
+
+        Args:
+        circuit (:class:`qibo.models.Circuit`): The Qibo circuit to be executed.
+        k (int): The index of the state prepared.
+        j (int): The index of the measurement basis.
+        nshots (int, optional): Number of shots to execute circuit with.
+    Returns:
+        numpy.float: Expectation value given by either Tr(Q_j rho_k) or Tr(Q_j O_l rho_k).
+    """
+
+    if j == 0:
+        return 1.0
+
+    nqubits = circuit.nqubits
+    if nqubits != 1 and nqubits != 2:
+        raise ValueError(
+            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
+        )
+
+    if backend is None:  # pragma: no cover
+        backend = GlobalBackend()
+
+    result = backend.execute_circuit(circuit, nshots=nshots)
+    observables = [symbols.I, symbols.Z, symbols.Z, symbols.Z]
+    observables_list = list(product(observables, repeat=nqubits))[j]
+    observable = 1
+    for q, obs in enumerate(observables_list):
+        if obs is not symbols.I:
+            observable *= obs(q)
+    observable = SymbolicHamiltonian(observable, nqubits=nqubits)
+    expectation_val = result.expectation_from_samples(observable)
+
+    return expectation_val
+
+
+def GST(nqubits=None, gate=None, nshots=int(1e4), noise_model=None, backend=None):
+    """Runs gate set tomography for a 1 or 2 qubit gate/circuit.
+
+        Args:
+        nshots (int, optional): Number of shots used in Gate Set Tomography.
+        gate (:class:`qibo.gates.abstract.Gate`, optional): The gate to perform gate set tomography on.
+            If gate=None, then gate set tomography will be performed for an empty circuit.
+        noise_model (:class:`qibo.noise.NoiseModel`, optional): Noise model applied
+            to simulate noisy computation.
+        backend (:class:`qibo.backends.abstract.Backend`, optional): Calculation engine.
+    Returns:
+        numpy.array: Numpy array with elements jk equivalent to either Tr(Q_j rho_k) or Tr(Q_j O_l rho_k).
+    """
+
+    # Check if class_qubit_gate corresponds to nqubits-gate.
+    if nqubits != 1 and nqubits != 2:
+        raise ValueError(
+            f"nqubits given as {nqubits}. nqubits needs to be either 1 or 2."
+        )
+
+    if backend is None:  # pragma: no cover
+        backend = GlobalBackend()
+    if gate is not None:
+        qb_gate = len(gate.qubits)
+        if nqubits != qb_gate:
+            raise ValueError(
+                f"Mismatched inputs: nqubits given as {nqubits}. {gate} is a {qb_gate}-qubit gate."
+            )
+        gate = gate.__class__(*list(range(qb_gate)), **gate.init_kwargs)
+
+    # GST for empty circuit or with gates
+    matrix_jk = np.zeros((4**nqubits, 4**nqubits))
+    for k in range(4**nqubits):
+        circ = prepare_states(k, nqubits)
+        if gate is not None:
```

Comment:
```suggestion
        circ, gates = prepare_states(k, nqubits)
        if gate is not None:
            if gate.name[-3::] == 'sp_0':
                circ_inv = Circuit(nqubits)
                circ_inv.add(gates[0])
                circ.add(circ_inv.invert())
            elif gate.name[-3::] == 'sp_1':
                circ_inv = Circuit(nqubits)
                circ_inv.add(gates[1])
                circ.add(circ_inv.invert())
            elif gate.name[-3::] == 'sp_t':
                circ.add(circ.invert())
```

Here I suggest some modifications for the `GST()` function to take into account state preparation. Reset channels are only available in simulation, so in order to make the method compatible with hardware, I am adding the inverse of the previous circuit to return to the `|0>` state. It is possible to indicate which qubit we want to prepare to zero in the gate name using `sp_0` (qubit 0), `sp_1` (qubit 1), `sp_t` (the two qubits). This way, we only need to build the list with the appropriate operations and indicate in which cases it is a channel that prepares a certain state.

Reply:
This is a very good idea. I did not know that reset channels don't work in hardware. In the spirit of your suggestion, i have created a new function called ```reset_register``` that takes in the ```circuit``` and ```sp_0```, ```sp_1```, or ```sp_t``` and outputs the inverse of that register or registers. That way we avoid modifying ```prepare_states(k, qubits)``` because ```prepare_states(k, qubits)``` is used to prepare the initial states of the quantum circuit for gate set tomography. I'll update the codes in the next version soon.